Add TRIM support - replaces #5925 and #7363

dweeezil opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Frequently-requested feature to help performance on SSDs and on various other SAN-like storage back-ends which support UNMAP/TRIM.
### Description
<!--- Describe your changes in detail -->
This is a port of Nextenta's TRIM support to the current master branch - post-device removal and post-device initialization.
### How Has This Been Tested?
New ZTS tests.  Also additions to ztest.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf commented on lib/libspl/include/sys/dkio.h:
```diff
@@ -476,6 +522,38 @@ typedef struct dk_updatefw_32 {
+	 * a couple unused pointers is too low to justify that risk.
+	 */
+	void			(*dfl_ck_func)(uint64_t, uint64_t, void *);
+	void			*dfl_ck_arg;
```
This constraint really doesn't apply to us, nothing on Linux will call this,  We should either drop this or wrap it and the assigned `zio_trim_check `functions with `__sun__` to make it clear this is dead code.

behlendorf commented on lib/libzfs/libzfs_pool.c:
```diff
@@ -2255,6 +2255,30 @@ zpool_scan(zpool_handle_t *zhp, pool_scan_func_t func, pool_scrub_cmd_t cmd)
+	zc.zc_cookie = (uintptr_t)&tci;
+
+	if (zfs_ioctl(hdl, ZFS_IOC_POOL_TRIM, &zc) == 0)
+		return (0);
```
This needs to be converted to a new in/out nvlist style ioctl and use `libzfs_core.c` and `lzc_ioctl`.

kpande commented on include/sys/metaslab_impl.h:
```diff
@@ -352,6 +353,11 @@ struct metaslab {
 	range_tree_t	*ms_allocatable;
 
+	range_tree_t	*ms_cur_ts;		/* currently prepared trims */
+	range_tree_t	*ms_prev_ts;		/* previous (aging) trims */
```
aging is american english and ageing is british english, both are considered correct

dweeezil commented:
Status update: This has just been refreshed with some fixes to the various deadlocks observed while running ztest (which have been left as separate commits).  Reviewers should look at 1eda73773 and the related follow-on commits in particular.  With these changes in place, zloop can now generally survive an overnight run.  I've not yet done any actual testing to make sure this plays well (i.e. actually works as opposed to simply not panicking) with allocation classes or non-trivial cases of device removal (other than those which are performed via ztest, which are plenty).  We (@behlendorf and myself) will begin addressing the review feedback shortly.  Also, some of the new ZTS tests need fixing.

dweeezil commented on module/zfs/metaslab.c:
```diff
@@ -1194,41 +1205,9 @@ metaslab_block_picker(avl_tree_t *t, uint64_t *cursor, uint64_t size,
- * ==========================================================================
- */
-static uint64_t
-metaslab_ff_alloc(metaslab_t *msp, uint64_t size)
```
If you're referring to the removal of `metaslab_ff_alloc()`, this appears to have been done as part of porting your openzfs/openzfs@8629fdcdf3ff4043e6772fc19af435e707d20523 patch from the OpenZFS PR.  The removal of this function does certainly seem like something that ought to be in its own standalone PR.

tonyhutter commented on man/man8/zpool.8:
```diff
@@ -2110,6 +2159,70 @@ again.
 Starts a resilver. If an existing resilver is already running it will be
 restarted from the beginning. Any drives that were scheduled for a deferred
 resilver will be added to the new one.
```
Add for newlines:
``` 
Starts a resilver. If an existing resilver is already running it will be
 restarted from the beginning. Any drives that were scheduled for a deferred
 resilver will be added to the new one.
+.It Xo
+.Nm
 .Cm trim
 .Op Fl p
 .Op Fl r Ar rate | Fl s
```

tonyhutter commented:
Include `atrim` and `mtrim` latency histograms in the man page
```
--- a/man/man8/zpool.8
+++ b/man/man8/zpool.8
@@ -1820,6 +1820,10 @@ Amount of time IO spent in asynchronous priority queues.  Does not include
 disk time.
 .Ar scrub :
 Amount of time IO spent in scrub queue. Does not include disk time.
+.Ar atrim :
+Amount of time IO spent doing a trim operation during an autotrim.
+.Ar mtrim :
+Amount of time IO spent doing a trim operation during a manual trim.
 .It Fl l
 Include average latency statistics:
 .Pp
```

tonyhutter commented:
You'll want to add the new `trim_start` and `trim_end` events to `man/man5/zfs-events.5`.

tonyhutter commented on tests/zfs-tests/tests/functional/trim/manualtrim_001_pos.ksh:
```diff
@@ -0,0 +1,100 @@
+#
+
+#
+# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
```
`Copyright 2007 Sun Microsystems` is this needed on new files?

kpande commented:
the CDDL header is copyrighted text

richardelling commented on tests/zfs-tests/tests/functional/cli_root/zpool_trim/cleanup.ksh:
```diff
@@ -0,0 +1,36 @@
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or http://www.opensolaris.org/os/licensing.
```
opensolaris.org is no more. Please refer to http://opensource.org/licenses/CDDL-1.0

tonyhutter commented on tests/runfiles/linux.run:
```diff
@@ -836,6 +841,11 @@ tags = ['functional', 'threadsappend']
+tests = ['autotrim_001_pos', 'autotrim_002_pos', 'manualtrim_001_pos',
+    'manualtrim_002_pos', 'manualtrim_003_pos', 'manualtrim_004_pos',
+    'manualtrim_005_pos']
+
```
Add `tags = ['functional', 'trim']`

RJVB commented:
> the CDDL header is copyrighted text

Not to make a fuss about this, but

- I think Tony's initial question was not about the CDDL header itself but on what's written underneath
- the CDDL header has a clear end marker so any copyright of that header should end there (?!)
- is it even legal to antedate a copyright on new code and attribute it to an organisation where you cannot even be an employe because it no longer exists?

dweeezil commented:
I'm going to be closing this PR.  @behlendorf has done some major restructuring of the original TRIM patch set to, among other things:

dweeezil commented:
I'm going to be closing this PR.  @behlendorf has been doing some major re-working of the original TRIM work to, among other things:
  - Harmonize its operation with that of device initialization so it's possible, for example, to TRIM individual vdevs.
  - Add per-vdev trim stats
  - Re-work the underlying discard mechanism to use a new ZIO type rather than the DKIOCFREE ioctl.
  - Utilize the new "xlat" vdev method from device initialization where appropriate.
  - Switch from the legacy ioctl interface to the new nvlist-style ioctl interface.
  - Plenty of other restructuring, in particular w.r.t. the management of the trim threads.

Expect a new PR to be posted soon.
